### ISSUE TITLE
Saving and loading tf_agents

### ISSUE THREAD

**Issue by robodhruv** [tensorflow/agents#235]
What is the right way to save/checkpoint a tf_agent and load it?
- `pickle` does not dump the agent and gives `can't pickle weakref objects`
- `tf.saved_model.save` successfully saves the agent but on loading, the signatures.keys() yields an empty list.

Tried this with a custom tf_agent as well as the standard DqnAgent from the tutorial Colab notebook.

**Comment by oars** [comment]
Take a look at the policy_saver for how to do saved_models, this is useful for deploying learned policies:
https://github.com/tensorflow/agents/blob/master/tf_agents/policies/policy_saver.py#L73

If you want to save and restore to train across multiple runs take a look at the checkpointer as used in the DQN example.:
https://github.com/tensorflow/agents/blob/master/tf_agents/agents/dqn/examples/v2/train_eval.py#L216

Feel free to re-open if you have any questions with these.

**Comment by robodhruv** [comment]
Thanks for your reply, @oars ! It is not clear how to restore a checkpoint: the `common.Checkpointer` does not have a restore() or similar method. can you point me to a relevant resource?

**Comment by oars** [comment]
See the use of `initialize_or_restore()`
https://github.com/tensorflow/agents/blob/master/tf_agents/agents/dqn/examples/v2/train_eval.py#L216

**Comment by cosmir17** [comment]
I am curious.. @oars Does it also load the policy network and agent information automatically? 

I have the following code:

```
q_net = q_network.QNetwork(
    nimble_quest_env.observation_spec(),
    nimble_quest_env.action_spec(),
    conv_layer_params=conv_layer_params,
    fc_layer_params=fc_layer_params)

optimizer = tf.compat.v1.train.AdamOptimizer(learning_rate=learning_rate)
#global_step = tf.compat.v1.train.get_or_create_global_step()
global_step = tf.compat.v1.train.get_global_step()

#########################################################################
agent = dqn_agent.DdqnAgent(
    nimble_quest_env.time_step_spec(),
    nimble_quest_env.action_spec(),
    q_network=q_net,
    optimizer=optimizer,
    td_errors_loss_fn=common.element_wise_squared_loss,
    train_step_counter=global_step)

agent.initialize()
##########################################################################

eval_policy = agent.policy
collect_policy = agent.collect_policy

replay_buffer = tf_uniform_replay_buffer.TFUniformReplayBuffer(
    data_spec=agent.collect_data_spec,
    batch_size=nimble_quest_env.batch_size,
    max_length=replay_buffer_max_length)

checkpoint_dir = os.path.join(tempdir, 'checkpoint')
train_checkpointer = common.Checkpointer(
    ckpt_dir=checkpoint_dir,
    max_to_keep=1,
    agent=agent,
    policy=agent.policy,
    replay_buffer=replay_buffer,
#global_step=global_step
)
policy_dir = os.path.join(tempdir, 'policy_ddqn')

train_checkpointer.initialize_or_restore()
global_step = tf.compat.v1.train.get_global_step()
tf_policy_saver = policy_saver.PolicySaver(agent.policy)
```

1. I couldn't have 'global_step' in `common.Checkpointer(...)` constructor when running second time and restoring from a checkpoint, because it complained that  `global_step` was None. That's why I had to run `global_step = tf.compat.v1.train.get_global_step()` again. Am I doing right?

2. After this, my python script resumed from the previous training step (step no. 121.. etc) but I got the message:
`function with signature contains input name(s) 0/step_type, 0/reward, 0/discount, 0/observation with unsupported characters which will be renamed to step_type, reward, discount, observation in the SavedModel`. Do you think if it's ok to ignore the message? Do I still have updated q-network weight from the checkpoint files?
